pkg/forward/forward.go

```diff
@@ -478,53 +478,73 @@ func (isdf *InterStepDataForward) writeToBuffers(ctx context.Context, messageToS
+		needRetry := false
+
+		for idx, msg := range messages {
+			if err := errs[idx]; err != nil {
+				// ATM there are no user defined errors during write, all are InternalErrors.
+
+				// using `strings.Contains` to check BufferFull error type is not a good practice for two reasons:
+				// 1. it's an O(n) operation which, if being called frequently, can introduce performance issue.
+				// 2. it assumes that all the buffer implementations implement err.Error() to use "Buffer full" as error message content.
+				// TODO - a better way could be to declare a BufferFullError type and use type check instead of checking error message.
+				needRetry = !(isdf.onFullActions[toBuffer.GetName()] == dfv1.DropAndAckLatest && strings.Contains(err.Error(), "Buffer full"))
+				if needRetry {
+					// we retry only failed messages
+					failedMessages = append(failedMessages, msg)
+					writeMessagesError.With(map[string]string{metrics.LabelVertex: isdf.vertexName, metrics.LabelPipeline: isdf.pipelineName, "buffer": toBuffer.GetName()}).Inc()
+					// a shutdown can break the blocking loop caused due to InternalErr
+					if ok, _ := isdf.IsShuttingDown(); ok {
+						platformError.With(map[string]string{metrics.LabelVertex: isdf.vertexName, metrics.LabelPipeline: isdf.pipelineName}).Inc()
+						return writeOffsets, fmt.Errorf("writeToBuffer failed, Stop called while stuck on an internal error with failed messages:%d, %v", len(failedMessages), errs)
+					}
+				} else {
+					// drop the message
+					dropBytes += float64(len(msg.Payload))
 				}
 			} else {
+				writeCount++
+				writeBytes += float64(len(msg.Payload))
 				// we support write offsets only for jetstream
 				if _writeOffsets != nil {
 					writeOffsets = append(writeOffsets, _writeOffsets[idx])
 				}
 			}
 		}
-		// set messages to failed for the retry
+
 		if needRetry {
-			isdf.opts.logger.Errorw("Retrying failed msgs", zap.Any("errors", errorArrayToMap(errs)))
+			isdf.opts.logger.Errorw("Retrying failed messages",
+				zap.Any("errors", errorArrayToMap(errs)),
+				zap.String("pipeline", isdf.pipelineName),
+				zap.String("vertex", isdf.vertexName),
+				zap.String("buffer", toBuffer.GetName()),
+			)
+			// set messages to failed for the retry
 			messages = failedMessages
 			// TODO: implement retry with backoff etc.
 			time.Sleep(isdf.opts.retryInterval)
-			goto retry
 		} else {
 			break
 		}
 	}
 
-	writeMessagesCount.With(map[string]string{metrics.LabelVertex: isdf.vertexName, metrics.LabelPipeline: isdf.pipelineName, "buffer": toBuffer.GetName()}).Add(float64(len(messages)))
-	writeBytesCount.With(map[string]string{metrics.LabelVertex: isdf.vertexName, metrics.LabelPipeline: isdf.pipelineName, "buffer": toBuffer.GetName()}).Add(totalBytes)
+	dropMessagesCount.With(map[string]string{metrics.LabelVertex: isdf.vertexName, metrics.LabelPipeline: isdf.pipelineName, "buffer": toBuffer.GetName()}).Add(float64(totalCount - writeCount))
```

Comment:
> 0 check?

Reply:
Synced up offline. I agree it's ok to emit metrics with value 0.